### PR TITLE
Allow passing explicit PrimaryAffinity

### PR DESCRIPTION
**Pull request by synarete**
Allow passing explicit primary-affinity upon OSD's deployment using
annotations mechanism (same as DeviceClass/InitialWeight). This may be
useful for cases where user wants to have more fine-grained control on
the overall (read) load of specific devices. Valid values are within
range [0, 1) (where 1 is the default).

As an example, consider the (real-life) case of a PV which resides on a
partition alongside OS partition. In such case, user can reduce I/O load
of this specific OSD by choosing explicit primary-affinity value.

Issue: https://issues.redhat.com/browse/KNIP-1616
BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1924946

Signed-off-by: Shachar Sharon <ssharon@redhat.com>

### CODE REVIEW

**Review by jarrpa — CHANGES_REQUESTED** [review]
Overall this seems to make sense. Two requests:
* Please provide at least a short description of what "primary affinity" means in the comments in `storagecluster_types.go`
* Split the generated CSV changes into their own commit

**Comment by synarete** [comment]
> Overall this seems to make sense. Two requests:
> 
> * Please provide at least a short description of what "primary affinity" means in the comments in `storagecluster_types.go`
> * Split the generated CSV changes into their own commit

@jarrpa Will fix soon and re-send the PR

File: api/v1/storagecluster_types.go [hunk 1]
```diff
@@ -163,6 +163,14 @@ type StorageDeviceSet struct {
 	// +optional
 	InitialWeight string `json:"initialWeight,omitempty"`
 
+	// PrimaryAffinity is an optional OSD primary-affinity value within the
+	// range [0,1). This value influence the way Ceph's CRUSH selection of
+	// primary OSDs. Lower value reduce performance bottlenecks (especially
+	// on read operations). If not set, default value is 1.
+	// +kubebuilder:validation:Pattern=`^0.[0-9]+$`
+	// +optional
```
Comment (jarrpa) [inline api/v1/storagecluster_types.go:171]:
```suggestion
	// PrimaryAffinity is an optional OSD primary-affinity value within the
	// range [0,1). This value influences Ceph's CRUSH selection of
	// primary OSDs. A lower value reduces the probability that the OSD will
	// be selected. If not set, default value is 1.
	// https://docs.ceph.com/en/latest/rados/operations/crush-map/#primary-affinity
	// +kubebuilder:validation:Pattern=`^0.[0-9]+$`
	// +optional
```

**Review by jarrpa — CHANGES_REQUESTED** [review]
Good split. Suggesting some edits to the additional comment. Your comment gave me a better understanding on what this does, but the language was a little unclear. A direct link to the documentation will probably be useful here, since anyone modifying this value is already in a place where they should know more about this.

File: api/v1/storagecluster_types.go (hunk 1 above)
Comment (synarete) [reply api/v1/storagecluster_types.go:171]: @jarrpa Should I re-submit my PR with those changes?

Comment (jarrpa) [reply api/v1/storagecluster_types.go:171]: Yes

**Review by jarrpa — APPROVED** [review]
/lgtm

**Comment by obnoxxx** [comment]
@synarete , @jarrpa - I think we need a backport of this to `release-4.8`.

**Comment by synarete** [comment]
> @synarete , @jarrpa - I think we need a backport of this to `release-4.8`.

@jarrpa I issued a PR (#1179)

**Pull request merged**